### PR TITLE
Junk Tractor Beam

### PR DESCRIPTION
**Pull request by hyperioo**
## About The Pull Request

Replace Junk Beacon by Junk Tractor Beam system.

* JTB is handled by a console in the Junk Beacon room. It opens a bluespace portal that allows free travel between the ship and the z-level of the junk field.

* Junk field has a 50% chance to have the asteroid property. If a junk field has this property, it will contain asteroids rich in minerals like the ones on the space ruins z-level.

* Junk field have a random affinity that affect the loot they contain. Affinities are Neutral, Serb, OneStar and IronHammer. For now this does not have any effect as I am waiting for the loot rework that will add biome-themed loot.

* Plenty of 5 by 5 chunks and a few 25 by 25 chunks are placed on the map, they contain the loot. They are randomly choosen in the chunk database and have a random orientation. So system is modular and it will be easy to add a heckin' lot of chunks to add variety and affinity-themed chunks.

* You better bring weapons, jonking is not easy anymore.

* With the JTB console you can choose to release the junk field, this will close the portal and clean the whole z-level. You cannot release the portal if there is an alive human or silicon in the junk field, as a security measure.

* Once released the junk tractor beams enter a 5 minute cooldown.

* After that, you will be able to choose among 3 random junk field. Once you turn on the field capture, it generates the z-level then opens a new portal.

* Junk field is of size 100 by 100 and surrounded by an energy field with a wrap-around effect.

* JTB pillars sprites by infrared_baron

**JTB room:**
jtb1

**Being greeted by local wildlife:**
jtb2

**Part of a 25x25 chunk:**
jtb3

## Why It's Good For The Game

**JONK**, is there anything more to say?

It makes jonking an actual challenge instead of just digging for half an hour using junk beacon to get all the gamer loot.

Please end my suffering the behaviour of that chunk loading function was hell ;-;

## Changelog
:cl: Hyperio
add: Add Junk Tractor Beam feature (junk field generation, JTB console, JTB chunk system, affinity system, asteroid generation)
fix: Prevents runtime in add_underlay function
/:cl:

### CODE REVIEW

**Comment by Voldirs** [comment]
Sounds sexy

**Comment by Wouju** [comment]
Decided to try this out before everyone else (also suggest posting a picture of the new area, unless you want me to)

how long does the cooling down last for by the way? 
image
or atleast a nudge on where to look

**Comment by Wouju** [comment]
> Decided to try this out before everyone else (also suggest posting a picture of the new area, unless you want me to)
> 
> how long does the cooling down last for by the way?
> image
> or atleast a nudge on where to look

speak of the devil, the moment i posted that the cooling down ended

**Comment by MLGTASTICa** [comment]
Does this mean that the junk beacon is going to have better loot overall ? perhaps deep-maint level of loot ? or just more challanging overall?

**Comment by Wouju** [comment]
> Does this mean that the junk beacon is going to have better loot overall ? perhaps deep-maint level of loot ? or just more challanging overall?

more chalenging i would say, comparable to deepmaint, as there's a fair amount of space carps when i was fucking around with the feature

**Comment by hyperioo** [comment]
@Wouju You can change the value of `var/beam_cooldown_time = 5 MINUTES ` to something like `10 SECONDS` if you want to do some quick tests.

@MLGTASTICa For now I would say mostly more challenging. The "better loot" will kinda be when the affinity system will be working because you will be able to choose which kind of theme you want for the loot.

**Comment by hyperioo** [comment]
Added pictures + a cooldown bar to the console screen.

File: code/game/objects/effects/portals.dm [hunk 1]
```diff
@@ -221,3 +221,16 @@ var/list/portal_cache = list()
...
+
+/*
+Portal used to move from the ship to the junk field generated by junk tractor beam
+Basically a portal without time limit and failchance
+*/
+/obj/effect/portal/jtb
+	name = "junk field portal"
+	desc = "A portal stabilized by heavy-duty machinery. It is safe to cross."
+	failchance = 0
+	entropy_value = 0
```
Comment (TheShown911) [inline code/game/objects/effects/portals.dm:233]: Entropy is fun, give it some entropy.

Comment (hyperioo) [reply code/game/objects/effects/portals.dm:233]: @TheShown911 @Wouju The problem is that if I put an entropy value you can be sure a bonobo player will spend 15 minutes going back an forth between the junk field and the ship just to make the entropy skyrocket.

Comment (TheShown911) [reply code/game/objects/effects/portals.dm:233]: People can already do that with moebius and IH teleporters.

**Review by Firefox13 — CHANGES_REQUESTED** [review]
Mapmerge your maps please

**Comment by hyperioo** [comment]
Here you go, fixed some stuff to have maps properly loaded in mapmerged format.

**Review by Firefox13 — APPROVED** [review]

**Pull request merged**